src/internal/plugins/controllers/alert-flash-content.ts

```diff
@@ -0,0 +1,91 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// this code should not depend on React typings, because it is portable between major versions
+interface RefShim<T> {
+  current: T | null;
+}
+
+export interface AlertFlashContentContext {
+  type: string;
+  headerRef: RefShim<HTMLElement>;
+  contentRef: RefShim<HTMLElement>;
+  actionsRef: RefShim<HTMLElement>;
+}
+
+export type ReplacementType = 'original' | 'remove' | ((container: HTMLElement) => void);
+export type ReplacementTypeSafe = 'original' | 'remove' | true;
+
+type RegisterReplacement = (type: 'header' | 'content', doReplacement: ReplacementType) => void;
+
+export interface AlertFlashContentResult {
+  update: () => void;
+  unmount: (containers: { replacementHeaderContainer: HTMLElement; replacementContentContainer: HTMLElement }) => void;
+}
+
+export interface AlertFlashContentConfig {
+  id: string;
+  runReplacer: (context: AlertFlashContentContext, registerReplacement: RegisterReplacement) => AlertFlashContentResult;
+}
+
+export interface AlertFlashContentRegistrationListener {
+  (provider?: AlertFlashContentConfig): void | (() => void);
+  cleanup?: void | (() => void);
+}
+
+export interface AlertFlashContentApiPublic {
+  registerContentReplacer(config: AlertFlashContentConfig): void;
+}
+
+export interface AlertFlashContentApiInternal {
+  clearRegisteredReplacer(): void;
+  onContentRegistered(listener: AlertFlashContentRegistrationListener): () => void;
+}
+
+export class AlertFlashContentController {
+  #listeners: Array<AlertFlashContentRegistrationListener> = [];
+  #cleanups = new Map<AlertFlashContentRegistrationListener, null | (() => void)>();
+  #provider?: AlertFlashContentConfig;
+
+  registerContentReplacer = (content: AlertFlashContentConfig) => {
+    if (this.#provider) {
+      console.warn(
+        `Cannot call \`registerContentReplacer\` with new provider: provider with id "${this.#provider.id}" already registered.`
+      );
+      return;
+    }
+    this.#provider = content;
+
+    // Notify existing components if registration happens after the components are rendered.
+    this.#listeners.forEach(listener => {
+      const cleanup = listener(this.#provider) ?? null;
+      this.#cleanups.set(listener, cleanup);
+    });
+  };
+
+  clearRegisteredReplacer = () => {
+    this.#provider = undefined;
+  };
+
+  onContentRegistered = (listener: AlertFlashContentRegistrationListener) => {
+    this.#listeners.push(listener);
+    this.#cleanups.set(listener, listener(this.#provider) ?? null);
```

Comment:
1. Should it really be called when provider is undefined?
2. The most important side effect of this function is now buried as a 2nd level call in this line. This needs to be more clear.

Reply:
Updated

Reply:
The point 1 is still not addressed

Reply:
Updated

Reply:
The listener accepts undefined provider, so no changes here.

Reply:
Even on the pages without any runtime API used, there is still going to be a side effect of this call.

I do not think it is safe, but if you think it is fine, handling possible impact of this decision is on you.

Reply:
Right. I updated the code so that when there is no provider the listener is accepted but not called: it is only called when there is provider.